ios/engine/KMEI/KeymanEngine/Classes/Manager.swift

```diff
@@ -0,0 +1,3018 @@
+        do {
+          let responseDict = try JSONSerialization.jsonObject(with: request.rawResponseData!,
+                                                              options: .mutableContainers) as? [AnyHashable: Any]
+          options = responseDict?[Key.options] as? [AnyHashable: Any] ?? [:]
+          let unsortedLanguages = (responseDict?[Key.languages] as? [AnyHashable: Any])?[Key.languages]
+            as? [[String: Any]]
+          languages = unsortedLanguages!.sorted { a, b -> Bool in
+            let aName = a[Key.name] as! String
+            let bName = b[Key.name] as! String
+            return aName.localizedCaseInsensitiveCompare(bName) == .orderedAscending
+          }
+
+          createKeyboardsInfo()
+          kmLog("Request completed -- \(languages.count) languages.", checkDebugPrinting: true)
+          currentRequest = nil
+
+          if let completionBlock = request.userInfo["completionBlock"] as? FetchKeyboardsBlock {
+            completionBlock(nil)
+          }
+
+          NotificationCenter.default.post(name: .keymanLanguagesUpdated, object: self)
+        } catch {
+          kmLog("Failed: \(error).", checkDebugPrinting: true)
+          let error = NSError(domain: "Keyman", code: 0,
+                              userInfo: [NSLocalizedDescriptionKey: error.localizedDescription])
+          NotificationCenter.default.post(name: .keymanLanguagesDownloadFailed, object: self,
+                                          userInfo: [NSUnderlyingErrorKey: error])
+        }
+      }
+    }
+  }
+
+  public func downloadRequestFailed(_ request: HTTPDownloadRequest) {
+    switch request.typeCode {
+    case .downloadFile:
+      downloadQueue = nil
+      let error = request.error ?? NSError(domain: "Keyman", code: 0, userInfo: nil)
+
+      kmLog("Keyboard download failed: \(error).", checkDebugPrinting: true)
+      let kbInfo = request.userInfo[Key.keyboardInfo]
+      let keyboard: [String: String]
+      if let kb = kbInfo as? [String: String] {
+        keyboard = kb
+      } else if let keyboards = kbInfo as? [[String: String]], !keyboards.isEmpty {
+        keyboard = keyboards[0]
+      } else {
+        kmLog("downloadRequestFailed(): Unexpected userInfo: \(String(describing: kbInfo))",
+          checkDebugPrinting: false)
+        return
+      }
+      let kbID = keyboard[Key.keyboardId]!
+      let kbVersion = keyboard[Key.keyboardVersion]
+      let isUpdate = (request.userInfo[Key.update] as? Int)! != 0
+
+      if !isUpdate {
+        let fileManager = FileManager.default
+        let fileName = request.url.lastPathComponent
+        if fileName.hasJavaScriptExtension {
+          if let kbPath = keyboardPath(forFilename: fileName, keyboardVersion: kbVersion) {
+            try? fileManager.removeItem(at: kbPath)
+          }
+        }
+        if fileName.hasFontExtension {
+          // TODO: Check why this doesn't match the error case in downloadRequestFinished().
```

Comment:
iOS networking - sometimes a 'failed' request doesn't directly call the `downloadRequestFailed` method.  If I recall correctly, the calls for `downloadRequestFailed` and `downloadRequestFinished` are mutually exclusive, so the code appears to be duplicated.

When I did the network library rework, I may not have fully optimized everything here in the rework/refactor from the previously-used outdated library; I believe I had the latter method call this method as a failsafe, since we interpret a non-200 response code as a failure even if iOS does not.

Looking again at it now, it does look safe for a cleanup/centralizing refactor.